contrib/firebaseio/src/main/java/io/FirebaseSource.java

```diff
@@ -0,0 +1,397 @@
+     * recieved.
+     * @param event {@link FirebaseEvent} to wrap
+     */
+    private void wrapEventWithInstant(final FirebaseEvent<T> event){
+      timestampRef.setValue(ServerValue.TIMESTAMP, new CompletionListener(){
+        @Override
+        public void onComplete(FirebaseError err, Firebase ref) {
+          if (err != null){
+            error = err;
+          }
+          ref.addListenerForSingleValueEvent(new ValueEventListener(){
+            @Override
+            public void onCancelled(FirebaseError err) {
+              error = err;
+            }
+            @Override
+            public void onDataChange(DataSnapshot timestamp) {
+              checkpoint.put(new Record<>(
+                  new Instant(timestamp.getValue(Long.class)),
+                  event));
+            }
+          });
+        }
+      });
+    }
+
+    @Override
+    public boolean advance() throws IOException {
+      if (error == null){
+        return this.checkpoint.advance();
+      }
+      throw error.toException();
+    }
+
+    @Override
+    public CheckpointMark getCheckpointMark() {
+      return checkpoint;
+    }
+
+    @Override
+    public byte[] getCurrentRecordId() throws NoSuchElementException {
+      return digest.digest(
+          ByteBuffer.allocate(12).putInt(
+              this.checkpoint.read().data.hashCode()).putLong(
+                  this.checkpoint.read().timestamp.getMillis()).array());
+    }
+
+    @Override
+    public UnboundedSource<FirebaseEvent<T>, ?> getCurrentSource() {
+      return this.source;
+    }
+
+    @Override
+    public Instant getWatermark() {
+      return this.checkpoint.minTimestamp();
+    }
+
+    @Override
+    public boolean start() throws IOException {
+      try {
+        this.source.auther.authenticate(this.ref);
+      } catch (Exception e) {
+        throw new IOException(e);
+      }
+      if (this.source.listenForChildEvents){
+        this.childListener = this.ref.addChildEventListener(this.childListener);
+      }
+      if (this.source.listenForValueEvents){
+        this.valueListener = this.ref.addValueEventListener(this.valueListener);
+      }
+      return this.advance();
+    }
+
+    @Override
+    public void close() throws IOException {
+      if (this.source.listenForChildEvents) {
+        this.ref.removeEventListener(this.childListener);
+      }
+      if (this.source.listenForValueEvents){
+        this.ref.removeEventListener(this.valueListener);
+      }
+      final CountDownLatch lock = new CountDownLatch(1);
+      timestampRef.removeValue(new CompletionListener(){
+
+        @Override
+        public void onComplete(FirebaseError err, Firebase arg1) {
+          lock.countDown();
+          if (err != null){
+            throw err.toException();
+          }
+        }
+      });
+      try {
+        lock.await();
+      } catch (InterruptedException e) {
+        e.printStackTrace();
+        throw new IOException(e);
```

Comment:
restore the thread interrupted state so the caller can see it. `Thread.currentThread().interrupt()`.